dNet/ClientPackets.cpp

```diff
@@ -284,17 +287,77 @@ void ClientPackets::HandleChatModerationRequest(const SystemAddress& sysAddr, Pa
 		receiver.push_back(static_cast<uint8_t>(character));
 	}
 
+	if (!receiver.empty()) {
+		if (std::string(receiver.c_str(), 4) == "[GM]") {
+			receiver = std::string(receiver.c_str() + 4, receiver.size() - 4);
+		}
+	}
+
 	stream.Read(messageLength);
 	for (uint32_t i = 0; i < messageLength; ++i) {
 		uint16_t character;
 		stream.Read(character);
 		message.push_back(static_cast<uint8_t>(character));
 	}
 
+	bool isBestFriend = false;
+
+	if (chatLevel == 1) {
+		// Private chat
+		LWOOBJID idOfReceiver = LWOOBJID_EMPTY;
+
+		{
+			sql::PreparedStatement* stmt = Database::CreatePreppedStmt("SELECT name FROM charinfo WHERE name = ?");
+			stmt->setString(1, receiver);
+
+			sql::ResultSet* res = stmt->executeQuery();
+
+			if (res->next()) {
+				idOfReceiver = res->getInt("id");
+			}
+		}
+
+		if (user->GetIsBestFriendMap().find(receiver) == user->GetIsBestFriendMap().end() && idOfReceiver != LWOOBJID_EMPTY) {
+			sql::PreparedStatement* stmt = Database::CreatePreppedStmt("SELECT * FROM friends WHERE (player_id = ? AND friend_id = ?) OR (player_id = ? AND friend_id = ?) LIMIT 1;");
+			stmt->setInt(1, entity->GetObjectID());
+			stmt->setInt(2, idOfReceiver);
+			stmt->setInt(3, idOfReceiver);
+			stmt->setInt(4, entity->GetObjectID());
+
+			sql::ResultSet* res = stmt->executeQuery();
+
+			if (res->next()) {
+				isBestFriend = res->getInt("best_friend") == 3;
+			}
+
+			if (isBestFriend) {
+				auto tmpBestFriendMap = user->GetIsBestFriendMap();
+				tmpBestFriendMap[receiver] = true;
+				user->SetIsBestFriendMap(tmpBestFriendMap);
+			}
+
+			delete res;
+			delete stmt;
+		}
+		else if (user->GetIsBestFriendMap().find(receiver) != user->GetIsBestFriendMap().end()) {
+			isBestFriend = true;
+		}
+	}
+
 	std::unordered_map<char, char> unacceptedItems;
-	bool bAllClean = Game::chatFilter->IsSentenceOkay(message, user->GetLastUsedChar()->GetGMLevel());
+	std::vector<std::string> segments = Game::chatFilter->IsSentenceOkay(message, entity->GetGMLevel(), !(isBestFriend && chatLevel == 1));
+
+	bool bAllClean = segments.empty();
+
 	if (!bAllClean) {
-		unacceptedItems.insert(std::make_pair((char)0, (char)message.length()));
+		for (const auto& item : segments) {
+			if (item == "") {
+				unacceptedItems.insert({ (char)0, (char)message.length()});
+				break;
+			}
+
+			unacceptedItems.insert({ message.find(item), item.length() });
```

Comment:
How does this work for multiple of the same word being rejected in the same sentence?  From my observations this would only ever highlight the first word and never highlight the later appearances of them.  

It also seems inefficient to use .find here when you parsed the string above already and know what word is rejected and where it is.  I would recommend changing how this gets the index of the first letter of the word such that it is done in isSentenceOkay above when we split the word up the first time.

In my mind something like the word number mapped to its length that way there isnt a need to .find() the word again and also supports rejecting multiple of the same word in a sentence.  Then instead of splitting the word up by space in isSentenceOkay, we split it up here and pass that to the method so we can use it as a reference here for what words to accept or reject.

Reply:
> How does this work for multiple of the same word being rejected in the same sentence? From my observations this would only ever highlight the first word and never highlight the later appearances of them.

This runs for every every flagged segment by the functions in the filter?

Reply:
> It also seems inefficient to use .find here when you parsed the string above already and know what word is rejected and where it is. I would recommend changing how this gets the index of the first letter of the word such that it is done in isSentenceOkay above when we split the word up the first time.

We don't already know where it is as far as I know?

Reply:
> > How does this work for multiple of the same word being rejected in the same sentence? From my observations this would only ever highlight the first word and never highlight the later appearances of them.
> 
> This runs for every every flagged segment by the functions in the filter?

.find returns the first occurrence of the word found and cares not about ones that are found already

Reply:
> .find returns the first occurrence of the word found and cares not about ones that are found already

Ah I see what you mean, I thought about that but thought wasn't worth implementing properly. If people can't gather a word is not ok in one place but ok in another then that is there problem. Will look into fixing the logic.,